src/engine/TransitivePath.cpp

```diff
@@ -215,361 +218,168 @@ size_t TransitivePath::getCostEstimate() {
+  if (isLeft) {
+    p->_lhs.treeAndCol = std::make_pair(leftOrRightOp, inputCol);
+  } else {
+    p->_rhs.treeAndCol = std::make_pair(leftOrRightOp, inputCol);
+  }
 
-      Id child = *pos;
-      ++pos;
-      size_t childDepth = positions.size();
-      if (childDepth <= maxDist && marks.count(child) == 0) {
-        // process the child
-        if (childDepth >= minDist) {
-          marks.insert(child);
-          if (rightIsVar || child == rightValue) {
-            size_t row = res.size();
-            res.emplace_back();
-            res(row, 0) = last_elem;
-            res(row, 1) = child;
-            for (size_t k = 2; k < resWidth + 1; k++) {
-              if (k - 2 < leftSideCol) {
-                res(row, k) = left(i, k - 2);
-              } else if (k - 2 > leftSideCol) {
-                res(row, k - 1) = left(i, k - 2);
-              }
-            }
-          }
-        }
-        // Add the child to the stack
-        MapIt it = edges.find(child);
-        if (it != edges.end()) {
-          positions.push_back(it->second->begin());
-          edgeCache.push_back(it->second);
-        }
+  // Note: The `variable` in the following structured binding is `const`, even
+  // if we bind by value. We deliberately make one unnecessary copy of the
+  // `variable` to keep the code simpler.
+  for (auto [variable, columnIndexWithType] :
+       leftOrRightOp->getVariableColumns()) {
+    ColumnIndex columnIndex = columnIndexWithType.columnIndex_;
+    if (columnIndex != inputCol) {
+      if (columnIndex > inputCol) {
+        columnIndexWithType.columnIndex_++;
+        p->_variableColumns[variable] = columnIndexWithType;
+      } else {
+        columnIndexWithType.columnIndex_ += 2;
+        p->_variableColumns[variable] = columnIndexWithType;
       }
+      p->_resultWidth++;
     }
-
-    if (i + 1 < left.size()) {
-      // reset everything for the next iteration
-      marks.clear();
-    }
-
-    last_result_end = res.size();
   }
-
-  *dynRes = std::move(res).toDynamic();
+  return p;
 }
 
 // _____________________________________________________________________________
-template <int SUB_WIDTH, int LEFT_WIDTH, int RES_WIDTH>
-void TransitivePath::computeTransitivePathRightBound(
-    IdTable* dynRes, const IdTable& dynSub, const IdTable& dynRight,
-    size_t rightSideCol, bool leftIsVar, size_t leftSubCol, size_t rightSubCol,
-    Id leftValue, size_t minDist, size_t maxDist, size_t resWidth) {
-  // Do the discovery from the right instead of the left.
-
-  using Map = ad_utility::HashMap<Id, std::shared_ptr<ad_utility::HashSet<Id>>>;
-  using MapIt = Map::iterator;
-
-  const IdTableView<SUB_WIDTH> sub = dynSub.asStaticView<SUB_WIDTH>();
-  const IdTableView<LEFT_WIDTH> right = dynRight.asStaticView<LEFT_WIDTH>();
-  IdTableStatic<RES_WIDTH> res = std::move(*dynRes).toStatic<RES_WIDTH>();
+bool TransitivePath::isBound() const {
+  return _lhs.isBound() && _rhs.isBound();
+}
```

Comment:
I don't quite understand yet if this can happen. I don't think we have an implementation for this?

Reply:
It probably shouldn't happen.